phoenix-core-client/src/main/java/org/apache/phoenix/iterate/OffsetResultIterator.java

```diff
@@ -32,32 +34,49 @@
  */
 public class OffsetResultIterator extends DelegateResultIterator {
     private int rowCount;
-    private int offset;
+    private final int offset;
+    private Tuple lastScannedTuple;
     private long pageSizeMs = Long.MAX_VALUE;
+    private boolean isIncompatibleClient = false;
 
     public OffsetResultIterator(ResultIterator delegate, Integer offset) {
         super(delegate);
         this.offset = offset == null ? -1 : offset;
+        this.lastScannedTuple = null;
     }
 
-    public OffsetResultIterator(ResultIterator delegate, Integer offset, long pageSizeMs) {
+    public OffsetResultIterator(ResultIterator delegate, Integer offset, long pageSizeMs,
+                                boolean isIncompatibleClient) {
         this(delegate, offset);
         this.pageSizeMs = pageSizeMs;
+        this.isIncompatibleClient = isIncompatibleClient;
     }
+
     @Override
     public Tuple next() throws SQLException {
+        long startTime = EnvironmentEdgeManager.currentTimeMillis();
         while (rowCount < offset) {
             Tuple tuple = super.next();
-            if (tuple == null) { return null; }
+            if (tuple == null) {
+                return null;
+            }
             if (tuple.size() == 0 || isDummy(tuple)) {
                 // while rowCount < offset absorb the dummy and call next on the underlying scanner
                 continue;
             }
             rowCount++;
-            // no page timeout check at this level because we cannot correctly resume
-            // scans for OFFSET queries until the offset is reached
+            lastScannedTuple = tuple;
+            if (!isIncompatibleClient) {
```

Comment:
what doesn an incompatibleclient mean? and in which cases will you have scenario where currentTime - startTime is more than LONG_MAX? and why do you need to do that check? 
Shouldn't you return immediately with dummyTuple when its an incompatible client?

Reply:
Incompatible client means that the client that does not understand new format of cells returned by this particular change and hence we expect server to return same old format of cells so that old (incompatible) client can continue to work.

> in which cases will you have scenario where currentTime - startTime is more than LONG_MAX?

LONG_MAX is only if paging size is not set. If it is set, then page timeout value will be used.